### ISSUE TITLE
xml2hash() merges text blocks not as expected

### ISSUE THREAD

**Issue by useperl** [yoreek/XML-Hash-XS#9]
CDATA blocks in an XML are just another way of representing text. If the text has no 'unsafe' characters like `<`, `>` etc, there should be no difference if the text is or is not enclosed in CDATA. This is what I would expect and XML::Simple::XMLin() works in this way, while XML::Hash::XS::xml2hash() does not:
```perl
#!/usr/bin/perl -w
use strict;
use 5.10.0;

use XML::Hash::XS;
use XML::Simple;
use Data::Dump qw(dump);

my $xml = "<a><![CDATA[Hello,]]><![CDATA[ world!\n]]></a>";

my $x = XML::Hash::XS::xml2hash($xml);
say "XML::Hash::XS: ", dump($x);

my $y = XML::Simple::XMLin($xml);
say "XML::Simple: ", dump($y);
```
```
XML::Hash::XS: ["Hello,", "world!"]
XML::Simple: "Hello, world!\n"
```
In case of XML::Hash::XS, we get an array of 2 elements instead of a single string. Parameter merge_text => 1 can solve this problem...
```perl
#!/usr/bin/perl -w
use strict;
use 5.10.0;

use XML::Hash::XS;
use XML::Simple;
use Data::Dump qw(dump);

my $xml = "<a><![CDATA[Hello,]]><![CDATA[ world!\n]]></a>";

my $x = XML::Hash::XS::xml2hash($xml, merge_text => 1);
say "XML::Hash::XS: ", dump($x);

my $y = XML::Simple::XMLin($xml);
say "XML::Simple: ", dump($y);
```
```
XML::Hash::XS: "Hello,world!"
XML::Simple: "Hello, world!\n"
```
... but here another, much worse, problem appears: even the pieces of text belonging to different tags are merged:
```perl
#!/usr/bin/perl -w
use strict;
use 5.10.0;

use XML::Hash::XS;
use XML::Simple;
use Data::Dump qw(dump);

my $xml = <<"__EOF_XML__";
<a>
    <value><![CDATA[Hello,]]></value>
    <value><![CDATA[ world!\n]]></value>
</a>
__EOF_XML__

my $x = XML::Hash::XS::xml2hash($xml, merge_text => 1);
say "XML::Hash::XS: ", dump($x);

my $y = XML::Simple::XMLin($xml);
say "XML::Simple: ", dump($y);
```
```
XML::Hash::XS: { value => "Hello,world!" }
XML::Simple: { value => ["Hello,", " world!\n"] }
```

In other words, with `merge_text => 0` you unexpectedly get an array out of a single tag like `<a><![CDATA[Hello,]]><![CDATA[ world!\n]]></a>` while with `merge_text => 1` you unexpectedly get texts from different tags "pasted" together.

As a result, xml2hash() gives wrong results no matter what value `merge_text` has.

**Comment by yoreek** [comment]
Hi! Thanks for your report.
To resolve this problem we can use "force_array=1" option together with "merge_text=1":

`
$ perl -Iblib/lib -Iblib/arch -MData::Dump -MXML::Hash::XS -e 'Data::Dump::dump xml2hash("<a><value><![CDATA[Hello,]]></value><value><![CDATA[ world!\n]]></value></a>", merge_text => 1, force_array => 1)'
`

`{ value => ["Hello,", " world!\n"] }`

You can set "force_array" option to specific elements as:
`force_array => [qw(value)]`

**Comment by yoreek** [comment]
Is this issue still actual for you?

**Comment by useperl** [comment]
Yes, this is still an issue for me. Sorry for keeping silence.
The workaround you offered can not unfortunately be a solution for the problem. I use your code to actually convert a real life XML, quite big (a bit over 1 Gigabyte), with millions of tags. I know the structure of the XML in advance, actually, I have a RelaxNG (RNG) schema of it. So I know exactly which nodes must be arrays. I use force_array for these nodes only (force_array => [qw(all the nodes which I know are arrays according to the schema)]).
On the other hand, I have no idea where I will encounter several CDATA blocks. Can be anywhere, in any node(s) which allow(s) text. I do not generate this XML myself, I get it from an external source, and I get a new XML, say, every hour. So one XML can have 2 CDATA blocks in some place, and in the next XML can have no CDATA blocks in that place, but 3 CDATA blocks in a completely different place.
So it is not at all possible to "turn on" force_array and merge_text for the nodes which can contain 2 CDATA blocks, because this means *every and each* tag in the XML. Moreover, I have to use force_array for those and only those nodes from which I really expect an array of values.

So for now I have to use a dumb workaround for this feature of your code: I traverse all the resulting hash (slowly! pure Perl, and not XS) and fix the nodes which I know should not be arrays, but your code has made them arrays just because they had 2 or more CDATA blocks.

I hope I made my problem clear, but tell me if I did not.

**Comment by yoreek** [comment]
I changed the logic of merging text nodes.
Can you please check the latest version 0.48?
